File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxm=1, minm=1, ans=INT_MIN;
        for(int i=0; i<(int)nums.size(); ++i){
            int curr=nums[i];
            if(curr==0){
                maxm=1, minm=1;
                ans=max(ans,0);
                continue;
            }
            int tmxm=maxm, tmnm=minm;
            if(curr>0){
                maxm=max(tmxm*curr, curr);
                minm=min(tmnm*curr, curr);
            }
            else{
                maxm=max(tmnm*curr, curr);
                minm=min(tmxm*curr, curr); 
            }
            ans=max(ans,maxm);
        }
        return ans;
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int n= nums.length;
        int pre= 0, suff= 0, maxm= Integer.MIN_VALUE;
        for(int i=0;i<n;i++){
            if(pre==0) pre=1;
            if(suff==0) suff=1;
            pre*= nums[i];
            suff*= nums[n-1-i];
            maxm= Math.max(maxm, Math.max(pre, suff));
        }
        return maxm;
    }
}